--- mesa-mesa-18.3.4/src/mesa/drivers/dri/meson.build.orig	2019-03-05 19:44:11.161362920 +0100
+++ mesa-mesa-18.3.4/src/mesa/drivers/dri/meson.build	2019-03-05 19:44:45.604494243 +0100
@@ -78,7 +78,7 @@
     filebase : 'dri',
     description : 'Direct Rendering Infrastructure',
     version : meson.project_version(),
-    variables : ['dridriverdir=${prefix}/' + dri_drivers_path],
+    variables : ['dridriverdir=' + dri_drivers_path],
     requires_private : dri_req_private,
   )
 endif
