diff -dur -x '*~' -x '*.orig' -x '*.rej' mesa-17.1.2.orig/src/mesa/drivers/dri/i965/brw_sampler_state.c mesa-17.1.2/src/mesa/drivers/dri/i965/brw_sampler_state.c
--- mesa-17.1.2.orig/src/mesa/drivers/dri/i965/brw_sampler_state.c	2017-06-05 23:09:09.000000000 +0200
+++ mesa-17.1.2/src/mesa/drivers/dri/i965/brw_sampler_state.c	2017-06-17 16:01:16.890619847 +0200
@@ -450,8 +450,10 @@
    /* Enable anisotropic filtering if desired. */
    unsigned max_anisotropy = BRW_ANISORATIO_2;
    if (sampler->MaxAnisotropy > 1.0f) {
-      min_filter = BRW_MAPFILTER_ANISOTROPIC;
-      mag_filter = BRW_MAPFILTER_ANISOTROPIC;
+      if (min_filter == BRW_MAPFILTER_LINEAR)
+         min_filter = BRW_MAPFILTER_ANISOTROPIC;
+      if (mag_filter == BRW_MAPFILTER_LINEAR)
+         mag_filter = BRW_MAPFILTER_ANISOTROPIC;
 
       if (sampler->MaxAnisotropy > 2.0f) {
 	 max_anisotropy =
