diff -durN Mesa-3.1.orig/configure.in Mesa-3.1/configure.in
--- Mesa-3.1.orig/configure.in	Fri Nov 19 17:04:54 1999
+++ Mesa-3.1/configure.in	Wed Apr 19 15:22:52 2000
@@ -280,8 +280,8 @@
     AM_CONDITIONAL(HAVE_GGI_GENKGI, test $build_ggi_genkgi_driver = yes)
   fi
 fi
-ggi_confdir=/usr/local/etc
-ggi_libdir=/usr/local/lib
+ggi_confdir=/etc
+ggi_libdir=/usr/lib
 AC_SUBST(ggi_confdir)
 AC_SUBST(ggi_libdir)
 AC_SUBST(GGI_CFLAGS)
