--- Mesa-9.1/src/gallium/targets/xa-vmwgfx/Makefile.am~	2013-02-22 23:28:22.000000000 +0100
+++ Mesa-9.1/src/gallium/targets/xa-vmwgfx/Makefile.am	2013-02-26 19:51:30.938891682 +0100
@@ -45,7 +45,8 @@
 	$(top_builddir)/src/gallium/winsys/svga/drm/libsvgadrm.la \
 	$(top_builddir)/src/gallium/drivers/svga/libsvga.la \
 	$(top_builddir)/src/gallium/drivers/trace/libtrace.la \
-	$(top_builddir)/src/gallium/drivers/rbug/librbug.la
+	$(top_builddir)/src/gallium/drivers/rbug/librbug.la \
+	$(LIBDRM_LIBS)
 
 nodist_EXTRA_libxatracker_la_SOURCES = dummy.cpp
 
--- Mesa-9.1/src/gbm/Makefile.am~	2013-02-16 02:32:00.000000000 +0100
+++ Mesa-9.1/src/gbm/Makefile.am	2013-02-26 19:53:54.295673416 +0100
@@ -19,7 +19,7 @@
 	main/backend.c \
 	main/common.c
 libgbm_la_LDFLAGS = -version-info 1:0
-libgbm_la_LIBADD = $(LIBUDEV_LIBS) $(LIBKMS_LIBS) $(DLOPEN_LIBS)
+libgbm_la_LIBADD = $(LIBUDEV_LIBS) $(LIBKMS_LIBS) $(DLOPEN_LIBS) $(LIBDRM_LIBS)
 
 if HAVE_EGL_PLATFORM_WAYLAND
 AM_CPPFLAGS = -DHAVE_WAYLAND_PLATFORM
--- Mesa-9.1/src/gallium/targets/xvmc-nouveau/Makefile.am.orig	2013-02-20 01:26:22.000000000 +0100
+++ Mesa-9.1/src/gallium/targets/xvmc-nouveau/Makefile.am	2013-02-27 20:42:00.199549898 +0100
@@ -48,6 +48,7 @@
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/xvmc/libxvmctracker.la \
 	$(top_builddir)/src/gallium/winsys/nouveau/drm/libnouveaudrm.la \
+	$(top_builddir)/src/gallium/drivers/nv30/libnv30.la \
 	$(top_builddir)/src/gallium/drivers/nv50/libnv50.la \
 	$(top_builddir)/src/gallium/drivers/nvc0/libnvc0.la \
 	$(top_builddir)/src/gallium/drivers/nouveau/libnouveau.la \
--- Mesa-9.1/src/mesa/libdricore/Makefile.am.orig	2013-02-20 01:26:22.000000000 +0100
+++ Mesa-9.1/src/mesa/libdricore/Makefile.am	2013-02-27 20:43:01.806214222 +0100
@@ -42,6 +42,7 @@
 libdricore@VERSION@_la_LDFLAGS = -version-number 1:0
 libdricore@VERSION@_la_LIBADD = \
         ../program/libdricore_program.la \
+	$(DLOPEN_LIBS)
         $()
 
 if HAVE_X86_ASM
