--- mesa-20090322/src/mesa/sparc/xform.S.BAD	2009-04-10 17:19:53.000000000 -0500
+++ mesa-20090322/src/mesa/sparc/xform.S	2009-04-10 17:20:57.000000000 -0500
@@ -17,7 +17,7 @@
 
 #include "sparc_matrix.h"
 
-#if defined(SVR4) || defined(__SVR4) || defined(__svr4__)
+#if defined(SVR4) || defined(__SVR4) || defined(__svr4__) || defined(__arch64__)
 	/* Solaris requires this for 64-bit. */
         .register %g2, #scratch
         .register %g3, #scratch
