--- Mesa-5.1/src/glw/Makefile.X11.orig	2003-12-18 00:55:46.000000000 +0100
+++ Mesa-5.1/src/glw/Makefile.X11	2003-12-22 23:37:39.820144152 +0100
@@ -15,10 +15,10 @@
 LIBDIR = $(TOP)/lib
 
 # Use these sources for Motif (and Xt) support.  Requires Motif header files.
-#SOURCES = GLwDrawA.c GLwMDrawA.c
+SOURCES = GLwDrawA.c GLwMDrawA.c
 
 # OR, use these sources for non-Motif (Xt only) build:
-SOURCES = GLwDrawA.c
+#SOURCES = GLwDrawA.c
 
 
 OBJECTS = $(SOURCES:.c=.o)
@@ -40,13 +40,13 @@
 clean:
 	-rm -f *.o *~
 
-targets: $(LIBDIR)/$(GLW_LIB) message
+targets: $(LIBDIR)/$(GLW_LIB)
 
 
 # Make the library
 $(LIBDIR)/$(GLW_LIB): $(OBJECTS)
 	$(TOP)/bin/mklib -o GLw -major $(MAJOR) -minor $(MINOR) \
-		-patch $(TINY) $(MKLIB_OPTIONS) -install $(LIBDIR) \
+		-patch $(TINY) -static -install $(LIBDIR) \
 		$(GLW_LIB_DEPS) $(OBJECTS)
 
 
--- Mesa-5.1/Make-config.orig	2003-12-18 00:55:43.000000000 +0100
+++ Mesa-5.1/Make-config	2003-12-22 23:36:24.521591272 +0100
@@ -414,12 +414,12 @@
 	"GL_LIB = libGL.so" \
 	"GLU_LIB = libGLU.so" \
 	"GLUT_LIB = libglut.so" \
-	"GLW_LIB = libGLw.so" \
+	"GLW_LIB = libGLw.a" \
 	"OSMESA_LIB = libOSMesa.so" \
 	"CC = gcc" \
 	"CXX = g++" \
-	"CFLAGS = -O3 -ansi -pedantic -fPIC -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE -DUSE_XSHM -DPTHREADS -I/usr/X11R6/include" \
-	"CXXFLAGS = -O3 -ansi -pedantic -fPIC -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE" \
+	"CFLAGS = $(OPTFLAGS) -ansi -pedantic -fPIC -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE -DUSE_XSHM -DPTHREADS -I/usr/X11R6/include" \
+	"CXXFLAGS = $(OPTFLAGS) -ansi -pedantic -fPIC -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE" \
 	"GLUT_CFLAGS = -fexceptions" \
 	"GL_LIB_DEPS = -L/usr/X11R6/lib -lX11 -lXext -lm -lpthread" \
 	"GLU_LIB_DEPS = -L$(TOP)/lib -lGL -lm" \
@@ -664,8 +664,8 @@
 	"OSMESA_LIB = libOSMesa.a" \
 	"CC = gcc" \
 	"CXX = g++" \
-	"CFLAGS = -O3 -ansi -pedantic -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE -DUSE_XSHM -DPTHREADS -I/usr/X11R6/include" \
-	"CXXFLAGS = -O3 -ansi -pedantic -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE" \
+	"CFLAGS = $(OPTFLAGS) -ansi -pedantic -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE -DUSE_XSHM -DPTHREADS -I/usr/X11R6/include" \
+	"CXXFLAGS = $(OPTFLAGS) -ansi -pedantic -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE" \
 	"GLUT_CFLAGS = -fexceptions" \
 	"GL_LIB_DEPS = " \
 	"OSMESA_LIB_DEPS = " \
