--- Mesa-9.1/src/gallium/targets/xa-vmwgfx/Makefile.am~	2013-02-22 23:28:22.000000000 +0100
+++ Mesa-9.1/src/gallium/targets/xa-vmwgfx/Makefile.am	2013-02-26 19:51:30.938891682 +0100
@@ -45,7 +45,8 @@
 	$(top_builddir)/src/gallium/winsys/svga/drm/libsvgadrm.la \
 	$(top_builddir)/src/gallium/drivers/svga/libsvga.la \
 	$(top_builddir)/src/gallium/drivers/trace/libtrace.la \
-	$(top_builddir)/src/gallium/drivers/rbug/librbug.la
+	$(top_builddir)/src/gallium/drivers/rbug/librbug.la \
+	$(LIBDRM_LIBS)
 
 nodist_EXTRA_libxatracker_la_SOURCES = dummy.cpp
 

