--- mesa-mesa-19.1.7/meson.build.orig	2019-09-17 12:31:43.000000000 +0200
+++ mesa-mesa-19.1.7/meson.build	2019-10-06 20:18:10.222928443 +0200
@@ -493,7 +493,8 @@
 with_gallium_xvmc = false
 if _xvmc != 'false'
   dep_xvmc = dependency('xvmc', version : '>= 1.0.6', required : _xvmc == 'true')
-  with_gallium_xvmc = dep_xvmc.found()
+  dep_xv = dependency('xv', required : _xvmc == 'true')
+  with_gallium_xvmc = dep_xvmc.found() and dep_xv.found()
 endif
 
 xvmc_drivers_path = get_option('xvmc-libs-path')
--- mesa-mesa-19.1.7/src/gallium/state_trackers/xvmc/meson.build.orig	2019-09-17 12:31:43.000000000 +0200
+++ mesa-mesa-19.1.7/src/gallium/state_trackers/xvmc/meson.build	2019-10-06 20:18:31.069482175 +0200
@@ -23,7 +23,7 @@
   files('attributes.c', 'block.c', 'context.c', 'surface.c', 'subpicture.c'),
   c_args : [c_vis_args],
   include_directories : [inc_common],
-  dependencies : [dep_xvmc, dep_x11_xcb, dep_xcb, dep_xcb_dri2, dep_xcb_dri3],
+  dependencies : [dep_xvmc, dep_xv, dep_x11_xcb, dep_xcb, dep_xcb_dri2, dep_xcb_dri3],
 )
 
 # These tests will not work without a working xvmc configuration. 
