e - Mesa-7.11/src/glsl/Makefile~	2011-02-16 18:31:15.419335627 +0100
+++ Mesa-7.11/src/glsl/Makefile	2011-02-16 18:44:53.149316402 +0100
@@ -208,6 +208,6 @@
 
 builtin_function.cpp: builtins/profiles/* builtins/ir/* builtins/tools/generate_builtins.py builtins/tools/texture_builtins.py builtin_compiler
 	@echo Regenerating builtin_function.cpp...
-	$(PYTHON2) $(PYTHON_FLAGS) builtins/tools/generate_builtins.py ./builtin_compiler > builtin_function.cpp
+	$(PYTHON2) $(PYTHON_FLAGS) builtins/tools/generate_builtins.py ./builtin_compiler > builtin_function.cpp || rm -f builtin_function.cpp
 
 -include depend
