--- Mesa-5.0.2/si-glu/libnurbs/interface/Makefile.am.orig	2003-05-15 15:58:16.000000000 +0200
+++ Mesa-5.0.2/si-glu/libnurbs/interface/Makefile.am	2003-10-05 23:46:01.438531760 +0200
@@ -8,7 +8,7 @@
 
 # Choose one of STANDALONE, LIBRARYBUILD, and GLBUILD. They are debugging
 # switches. grep LIBRARYBUILD `find .. -name '*.h'` for details.
-CXXFLAGS += -DLIBRARYBUILD -DNDEBUG
+AM_CXXFLAGS = -DLIBRARYBUILD -DNDEBUG
 
 include libNIFac_la_SOURCES
 
--- Mesa-5.0.2/si-glu/libnurbs/internals/Makefile.am.orig	2003-05-15 15:58:16.000000000 +0200
+++ Mesa-5.0.2/si-glu/libnurbs/internals/Makefile.am	2003-10-05 23:46:20.456640568 +0200
@@ -8,7 +8,7 @@
 
 # Choose one of STANDALONE, LIBRARYBUILD, and GLBUILD. They are debugging
 # switches. grep LIBRARYBUILD `find .. -name '*.h'` for details.
-CXXFLAGS += -DLIBRARYBUILD -DNDEBUG
+AM_CXXFLAGS = -DLIBRARYBUILD -DNDEBUG
 
 include libNInt_la_SOURCES
 
--- Mesa-5.0.2/si-glu/libnurbs/nurbtess/Makefile.am.orig	2003-05-15 15:58:16.000000000 +0200
+++ Mesa-5.0.2/si-glu/libnurbs/nurbtess/Makefile.am	2003-10-05 23:46:29.455272568 +0200
@@ -8,7 +8,7 @@
 
 # Choose one of STANDALONE, LIBRARYBUILD, and GLBUILD. They are debugging
 # switches. grep LIBRARYBUILD `find .. -name '*.h'` for details.
-CFLAGS += -DLIBRARYBUILD -DNDEBUG
+AM_CFLAGS = -DLIBRARYBUILD -DNDEBUG
 
 include libNTess_la_SOURCES
 
--- Mesa-5.0.2/src-glut/Makefile.am.orig	2003-05-15 15:58:20.000000000 +0200
+++ Mesa-5.0.2/src-glut/Makefile.am	2003-10-05 23:47:38.298806760 +0200
@@ -9,7 +9,6 @@
 EXTRA_PROGRAMS = capturexfont
 
 libglut_la_LDFLAGS = -version-info @LIBGLUT_VERSION_INFO@ -L../@GLU_DIR@
-libglut_la_LIBADD = -lGLU
 libglut_la_SOURCES = \
 	glutwin32.h \
 	win32_glx.h \
@@ -67,7 +66,7 @@
 	layerutil.h \
 	stroke.h
 
-libglut_la_LIBADD = $(X_LIBS) $(X_PRE_LIBS) $(X_LIBADD)
+libglut_la_LIBADD = $(X_LIBS) $(X_PRE_LIBS) $(X_LIBADD) -lGLU
 capturexfont_LDFLAGS = $(X_LIBS) $(X_PRE_LIBS) $(X_LIBADD)
 
 EXTRA_DIST = descript.mms
--- Mesa-5.0.2/demos/Makefile.am.orig	2003-05-15 15:58:14.000000000 +0200
+++ Mesa-5.0.2/demos/Makefile.am	2003-10-05 23:48:00.977359096 +0200
@@ -44,7 +44,7 @@
 	tunnel2 \
 	winpos
 
-#	osdemo \
+#	osdemo
 
 
 CLEANFILES = test.ppm
--- Mesa-5.0.2/ggi/ggiglut/Makefile.am.orig	2003-05-15 15:58:14.000000000 +0200
+++ Mesa-5.0.2/ggi/ggiglut/Makefile.am	2003-10-05 23:48:27.849273944 +0200
@@ -3,7 +3,7 @@
 if HAVE_GGI
   lib_LTLIBRARIES = libglut.la
   libglut_la_SOURCES = ggiglut.c debug.h
-  libglut_la_LDFLAGS = -version-info $(LIBGLUT_VERSION)
+  libglut_la_LDFLAGS = -version-info $(LIBGLUT_VERSION_INFO)
 endif
 
 ggiincludedir = $(includedir)/ggi
--- Mesa-5.0.2/src/GGI/Makefile.am.orig	2003-05-15 15:58:19.000000000 +0200
+++ Mesa-5.0.2/src/GGI/Makefile.am	2003-10-05 23:51:19.580166880 +0200
@@ -7,7 +7,7 @@
 ggimesaconfdatadir = ${sysconfdir}/ggi
 ggimesaconfdata_DATA = ggimesa.conf
 
-INCLUDES += -I$(top_srcdir)/include -I$(top_srcdir)/src/GGI/include \
+INCLUDES = -I$(top_srcdir)/include -I$(top_srcdir)/src/GGI/include \
 	-I$(top_srcdir)/src $(GGI_CFLAGS)
 
 DEFS += -DGGIMESACONFFILE=\"$(ggimesaconfdatadir)/ggimesa.conf\"
--- Mesa-5.0.2/configure.in.orig	2003-10-05 23:51:56.000000000 +0200
+++ Mesa-5.0.2/configure.in	2003-10-05 23:54:09.250373088 +0200
@@ -461,6 +461,9 @@
     
   AM_CONDITIONAL(HAVE_GGI_FBDEV, test $build_ggi_fbdev_target = yes)
   AM_CONDITIONAL(HAVE_GGI_GENKGI, test $build_ggi_genkgi_driver = yes)
+else
+  AM_CONDITIONAL(HAVE_GGI_FBDEV, false)
+  AM_CONDITIONAL(HAVE_GGI_GENKGI, false)
 fi
 ggi_confdir=$prefix/etc
 ggi_libdir=$prefix/lib
--- Mesa-5.0.2/src/X86/Makefile.am.orig	2003-05-15 15:58:20.000000000 +0200
+++ Mesa-5.0.2/src/X86/Makefile.am	2003-10-06 00:07:21.525928896 +0200
@@ -77,6 +77,7 @@
 libMesaX86_la_SOURCES = $(x86_sources) $(mmx_sources) $(x3dnow_sources) $(sse_sources)
 
 INCLUDES = -I$(top_srcdir)/include -I$(top_srcdir)/src $(CFLAGS_X86) $(CFLAGS_MMX) $(CFLAGS_3DNOW) $(CFLAGS_SSE)
+AM_CCASFLAGS = $(INCLUDES)
 
 EXTRA_DIST = $(x3dnow_files) $(mmx_files) $(sse_files)
 
