--- Mesa-6.4.1/src/glx/x11/dri_glx.c.orig	2005-12-20 12:29:38.000000000 +0000
+++ Mesa-6.4.1/src/glx/x11/dri_glx.c	2005-12-20 12:30:44.000000000 +0000
@@ -58,7 +58,7 @@
 
 #ifndef DEFAULT_DRIVER_DIR
 /* this is normally defined in Mesa/configs/default with DRI_DRIVER_SEARCH_PATH */
-#define DEFAULT_DRIVER_DIR "/usr/X11R6/lib/modules/dri"
+#define DEFAULT_DRIVER_DIR "/usr/lib/xorg/modules/dri"
 #endif
 
 static __DRIdriver *Drivers = NULL;

