diff -ur Mesa-7.1.orig/src/mesa/drivers/dri/common/dri_util.h Mesa-7.1/src/mesa/drivers/dri/common/dri_util.h
--- Mesa-7.1.orig/src/mesa/drivers/dri/common/dri_util.h	2008-09-06 23:18:43.000000000 +0200
+++ Mesa-7.1/src/mesa/drivers/dri/common/dri_util.h	2008-09-06 23:23:05.000000000 +0200
@@ -53,6 +53,9 @@
 #include <drm.h>
 #include <drm_sarea.h>
 #include <xf86drm.h>
+#ifdef TTM_API
+#include <xf86mm.h>
+#endif
 #include "glheader.h"
 #include "GL/internal/glcore.h"
 #include "GL/internal/dri_interface.h"
--- Mesa-7.1/src/mesa/drivers/dri/intel/intel_bufmgr_ttm.c.orig	2008-09-06 23:28:18.000000000 +0200
+++ Mesa-7.1/src/mesa/drivers/dri/intel/intel_bufmgr_ttm.c	2008-09-06 23:26:50.000000000 +0200
@@ -51,6 +51,7 @@
 
 #include "intel_bufmgr_ttm.h"
 #ifdef TTM_API
+#include <xf86mm.h>
 
 #define DBG(...) do {					\
    if (bufmgr_ttm->bufmgr.debug)			\
