--- Mesa-3.4.2/widgets-sgi/Makefile.X11.orig	Thu Jun 15 20:46:09 2000
+++ Mesa-3.4.2/widgets-sgi/Makefile.X11	Sun May 27 14:37:57 2001
@@ -16,9 +16,9 @@
 INCDIRS = -I../include -I/usr/include/Motif1.2 -I/usr/X11R6/include
 LIBDIR = ../lib
 
 # Compile GLwMDrawA.c only if you have Motif headers!
-#SOURCES = GLwDrawA.c GLwMDrawA.c
-SOURCES = GLwDrawA.c
+SOURCES = GLwDrawA.c GLwMDrawA.c
+#SOURCES = GLwDrawA.c
 
 
 OBJECTS = $(SOURCES:.c=.o)
@@ -40,14 +41,12 @@
 clean:
 	-rm *.o *~
 
-targets: $(LIBDIR)/$(GLW_LIB) message
+targets: $(LIBDIR)/$(GLW_LIB)
 
 
 # Make the library
 $(LIBDIR)/$(GLW_LIB): $(OBJECTS)
 	$(MAKELIB) $(GLW_LIB) $(MAJOR) $(MINOR) $(TINY) $(OBJECTS)
-	mv $(GLW_LIB)* $(LIBDIR)
-	-mv *.a $(LIBDIR)
 
 
 message:
--- Mesa-4.0.1/Make-config.orig	Mon May 14 21:11:32 2001
+++ Mesa-4.0.1/Make-config	Sun May 27 14:31:38 2001
@@ -453,9 +453,9 @@
 	"GLUT_LIB = libglut.so" \
-	"GLW_LIB = libGLw.so" \
+	"GLW_LIB = libGLw.a" \
 	"OSMESA_LIB = libOSMesa.so" \
 	"CC = gcc" \
	"CPLUSPLUS = g++" \
-	"CFLAGS = -O3 -ansi -pedantic -fPIC -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE -DUSE_XSHM -DPTHREADS -I/usr/X11R6/include" \
+	"CFLAGS = $(OPTFLAGS) -ansi -pedantic -fPIC -ffast-math -D_POSIX_SOURCE -D_POSIX_C_SOURCE=199309L -D_SVID_SOURCE -D_BSD_SOURCE -DUSE_XSHM -DPTHREADS -I/usr/X11R6/include" \
 	"CCFLAGS = $(CFLAGS)" \
 	"MAKELIB = ../bin/mklib.linux" \
 	"GL_LIB_DEPS = -L/usr/X11R6/lib -lX11 -lXext -lm -lpthread" \
