diff -dur mesa-17.2.1.orig/src/gallium/drivers/swr/Makefile.am mesa-17.2.1/src/gallium/drivers/swr/Makefile.am
--- mesa-17.2.1.orig/src/gallium/drivers/swr/Makefile.am	2017-09-18 01:03:22.000000000 +0200
+++ mesa-17.2.1/src/gallium/drivers/swr/Makefile.am	2017-09-18 10:31:22.000000000 +0200
@@ -236,6 +236,7 @@
 COMMON_LDFLAGS = \
 	-shared \
 	-module \
+	-avoid-version \
 	-no-undefined \
 	$(GC_SECTIONS) \
 	$(LD_NO_UNDEFINED)
