diff -urN Mesa-4.0.1.org/src/GGI/ggimesa.c Mesa-4.0.1/src/GGI/ggimesa.c
--- Mesa-4.0.1.org/src/GGI/ggimesa.c	Thu Jan  3 17:34:03 2002
+++ Mesa-4.0.1/src/GGI/ggimesa.c	Thu Jan  3 18:54:42 2002
@@ -46,7 +46,7 @@
 static void *_ggimesaConfigHandle;
 
 /* FIXME: These should really be defined in the make system using -Dxxx */
-#define GGIMESACONFFILE "/usr/local/etc/ggi/ggimesa.conf"
+#define GGIMESACONFFILE "/etc/ggi/ggimesa.conf"
 #define GGIMESATAGLEN 0
 static char ggimesaconfstub[512] = GGIMESACONFFILE;
 static char *ggimesaconffile = ggimesaconfstub + GGIMESATAGLEN;
diff -urN Mesa-4.0.1.org/src/config.c Mesa-4.0.1/src/config.c
--- Mesa-4.0.1.org/src/config.c	Thu Jan  3 17:34:01 2002
+++ Mesa-4.0.1/src/config.c	Thu Jan  3 18:54:51 2002
@@ -405,7 +405,7 @@
 #if defined(__WIN32__) || defined(__MSDOS__)
    const char *filename = "mesa.cnf";
 #else
-   const char *filename = "/etc/mesa.conf";
+   const char *filename = "/etc/X11/mesa.conf"; 
 #endif
    FILE *file;
    struct cnode *list;
