diff -urN Mesa.org/src/mesa/drivers/dri/r300/r300_state.c Mesa/src/mesa/drivers/dri/r300/r300_state.c
--- Mesa.org/src/mesa/drivers/dri/r300/r300_state.c	2006-04-11 13:41:11.000000000 +0200
+++ Mesa/src/mesa/drivers/dri/r300/r300_state.c	2006-04-11 23:12:50.622594500 +0200
@@ -1275,7 +1275,7 @@
 	DECLARE_RENDERINPUTS(index_bitset);      /* !hw_tcl_on */
 };
 
-static GLboolean r300_outputs_written_test (r300_outputs_written *ow, GLuint vp_result,
+static GLboolean r300_outputs_written_test (union r300_outputs_written *ow, GLuint vp_result,
                                             GLuint tnl_attrib)
 {
 	if (hw_tcl_on)

