--- Mesa-8.0.1/src/egl/wayland/wayland-drm/Makefile.orig	2012-02-15 00:44:00.000000000 +0100
+++ Mesa-8.0.1/src/egl/wayland/wayland-drm/Makefile	2012-03-11 20:50:16.884057197 +0100
@@ -17,8 +17,8 @@
 
 # Generate protocol sources
 prefix=$(shell $(PKG_CONFIG) --variable=prefix wayland-server)
-exec_prefx=$(shell $(PKG_CONFIG) --variable=exec_prefix wayland-server)
-wayland_protocoldir = $(PWD)/protocol
+exec_prefix=$(shell $(PKG_CONFIG) --variable=exec_prefix wayland-server)
+wayland_protocoldir = $(TOP)/src/egl/wayland/wayland-drm/protocol
 wayland_scanner=$(exec_prefix)/bin/wayland-scanner
 
 default: depend libwayland-drm.a $(GEN_SOURCES) $(GEN_HEADERS)
