diff -dur -x '*~' -x '*.orig' -x '*.rej' mesa-17.1.6.orig/src/gallium/drivers/swr/Makefile.am mesa-17.1.6/src/gallium/drivers/swr/Makefile.am
--- mesa-17.1.6.orig/src/gallium/drivers/swr/Makefile.am	2017-08-07 14:04:30.000000000 +0200
+++ mesa-17.1.6/src/gallium/drivers/swr/Makefile.am	2017-08-07 21:03:24.230258941 +0200
@@ -159,11 +159,13 @@
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/mesa/libmesagallium.la \
 	$(LLVM_LIBS) \
-	$(PTHREAD_LIBS)
+	$(PTHREAD_LIBS) \
+	-ldl
 
 COMMON_LDFLAGS = \
 	-shared \
 	-module \
+	-avoid-version \
 	-no-undefined \
 	$(GC_SECTIONS) \
 	$(NO_UNDEFINED) \
