diff -dur mesa-mesa-18.3.0.orig/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp mesa-mesa-18.3.0/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp
--- mesa-mesa-18.3.0.orig/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp	2018-12-07 12:20:00.000000000 +0100
+++ mesa-mesa-18.3.0/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp	2018-12-11 12:52:45.000000000 +0100
@@ -745,9 +745,6 @@
 {
    if (!i)
       i = new_Instruction(pol.context(), op, dType);
-#ifndef NDEBUG // non-conformant assert, so this is required
-   assert(typeid(*i) == typeid(*this));
-#endif
 
    pol.set<Instruction>(this, i);
 
