--- mesa-24.2.0/src/util/blake3/meson.build.orig	2024-08-14 18:37:13.000000000 +0200
+++ mesa-24.2.0/src/util/blake3/meson.build	2024-08-19 14:45:23.812609378 +0200
@@ -24,6 +24,8 @@
     else
       files_blake3 += ['blake3_sse2_x86-64_windows_gnu.S', 'blake3_sse41_x86-64_windows_gnu.S', 'blake3_avx2_x86-64_windows_gnu.S', 'blake3_avx512_x86-64_windows_gnu.S']
     endif
+  elif meson.get_compiler('c').has_define('__ILP32__')
+    blake3_defs += blake3_x86_no_simd_defs
   else
     files_blake3 += ['blake3_sse2_x86-64_unix.S', 'blake3_sse41_x86-64_unix.S', 'blake3_avx2_x86-64_unix.S', 'blake3_avx512_x86-64_unix.S']
   endif
