--- mesa-11.0.3/include/D3D9/d3d9types.h~	2015-10-02 10:45:51.000000000 +0200
+++ mesa-11.0.3/include/D3D9/d3d9types.h	2015-10-11 23:43:57.442310720 +0200
@@ -178,7 +178,7 @@
 #undef WINAPI
 #endif /* WINAPI*/
 
-#if defined(__x86_64__) || defined(_M_X64)
+#if (defined(__x86_64__) || defined(_M_X64)) && !defined(__ILP32__)
 #define WINAPI __attribute__((ms_abi))
 #else /* x86_64 */
 #define WINAPI __attribute__((__stdcall__))
